### DISCUSSION TITLE
Flutter on macOS doesn't work without hardwre acceleration

### DISCUSSION THREAD

**Discussion by Nadahar** [rustdesk/rustdesk#7651] [General]
I've struggled a lot to get the Flutter version of RustDesk working on my macOS VM. In the end I found the following when manually launching `rustdesk` from inside the "app" (Show Package Contents -> Contents -> MacOS -> rustdesk) which gave console output:
```
2024-04-05 10:19:02.237 RustDesk[3496:45964] Could not acquire Metal device.
2024-04-05 10:19:02.273 RustDesk[3496:45964] Unable to create FlutterView; no MTLDevice or MTLCommandQueue available.
2024-04-05 10:19:02.275 RustDesk[3496:45964] Unable to create FlutterView; no MTLDevice or MTLCommandQueue available.
2024-04-05 10:19:02.275 RustDesk[3496:45964] Unable to create FlutterView; no MTLDevice or MTLCommandQueue available.
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:19:02.498 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:19:39.001 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:19:56.692 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:19:56.704 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:19:56.712 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
embedder.cc (2450): 'FlutterEngineSendPlatformMessage' returned 'kInvalidArguments'. Invalid engine handle.
2024-04-05 10:20:00.373 RustDesk[3496:45964] Failed to send message to Flutter engine on channel 'flutter/lifecycle' (2).
2024-04-05 10:20:00.520 RustDesk[3496:46447] Persistent UI failed to open file file:///Users/<username>/Library/Saved%20Application%20State/com.carriez.rustdesk.savedState/window_1.data: No such file or directory (2)
```

I then found this: https://github.com/flutter/flutter/issues/108304

For those that don't know, "Metal" is Applish for "hardware acceleration", primarily 3D. It's their alternative to Microsoft DirectX and to the open source OpenGL. Since it is proprietary Apple only a few GPUs support this, and I don't think any software emulation exists. MacOS refuses to install if it can't find a "Metal-capable" GPU, but once you get past that, macOS will work just fine. You might experience a few graphical glitches in a some applications, but generally it's perfectly usable. Since "Metal" can't be emulated, most VMs don't support it - since that would mean that you'd have to install one of the few, expensive, "Metal-capable" GPU's and then dedicate it to said VM.

When Flutter chooses to remove support for OpenGL, "to save binary size", what they're actually doing is excluding all macOS versions before 10.14 and 99% of all VMs. As I see it, this is a big red flag when it comes to relying on Flutter. Flutter is thus "broken" for macOS since 3.7. I'm pretty sure they will do the same on other platforms and (intentionally) break things that Silicon Valley no longer want people to use.

I really hope you don't remove the Sciter implementation from RustDesk, because to me, and to many others I suspect, a remote control solution is only really viable if you can use it on all the systems you need to control. If you can only use it on 90%, and have to find another solution for the remaining, it's not really a workable solution.

**Comment by Nadahar** [comment]
Btw: I'm not suggesting that you have dropped Sciter - I'm just hoping that you're not planning to either (I think I saw it marked as deprecated somewhere).

I just built RustDesk on Fedora 39 as well, and while the Sciter version works well (after a couple of correction in `rpm.spec`), the Flutter version doesn't even open - no errors or nothing. So, I thought maybe there is something with my build, so I downloaded 1.2.3, both Flatpak and AppImage. The Flatpak version won't start because it complains over `org.freedesktop.Platform` missing, the AppImage does the same as my build - nothing. So, I think I'll just forget about the Flutter builds and live a happier life 😉 

edit: I discovered that Flutter build for RHEL simply isn't implemented in the build script, which is why the build was defective. I've actually managed to get the Flutter version to run on Fedora 39 after fixing the bridge generator not finding `stdbool.h` and running the Flutter build manually.

**Reply by matthewpipes** [reply]
> Btw: I'm not suggesting that you have dropped Sciter - I'm just hoping that you're not planning to either (I think I saw it marked as deprecated somewhere).
> 
> I just built RustDesk on Fedora 39 as well, and while the Sciter version works well (after a couple of correction in `rpm.spec`), the Flutter version doesn't even open - no errors or nothing. So, I thought maybe there is something with my build, so I downloaded 1.2.3, both Flatpak and AppImage. The Flatpak version won't start because it complains over `org.freedesktop.Platform` missing, the AppImage does the same as my build - nothing. So, I think I'll just forget about the Flutter builds and live a happier life 😉
> 
> edit: I discovered that Flutter build for RHEL simply isn't implemented in the build script, which is why the build was defective. I've actually managed to get the Flutter version to run on Fedora 39 after fixing the bridge generator not finding `stdbool.h` and running the Flutter build manually.

Thanks for this thread, I was thinking all hope is lost of controlling my MacOS 15 VM with RustDesk. Where did you get the sciter version of RustDesk for macOS? Did you build it yourself?

**Comment by open-trade** [comment]
I would suggest you submit to google Flutter team about this issue without GPU.

**Comment by Nadahar** [comment]
> Thanks for this thread, I was thinking all hope is lost of controlling my MacOS 15 VM with RustDesk. Where did you get the sciter version of RustDesk for macOS? Did you build it yourself?

Yes, I built it myself. This was all a year ago now though, so I don't remember the details.

**Comment by fragtion** [comment]
I have the same problem
